### DISCUSSION TITLE
Do tests fail when unable to load assets?

### DISCUSSION THREAD

**Discussion by Calpoog** [leancodepl/patrol#1692] [Q&A]
I've got an app package that's an app only in order to do integration testing. The contents of my package are actually meant to be consumed by other apps. Because of that, all assets are referenced with package paths, aka "packages/my_packages/asset/path/file".

I'm seeing briefly when the app pops up during a patrol test that the assets are showing as having failed to load. There was historically a Flutter bug that package paths didn't work in tests within that same package – however, it is supposedly fixed:
https://github.com/flutter/flutter/pull/111350. Unfortunately, due to Patrol's lack of exception/log capturing all I know is that the test failed to run with "multiple exceptions".

Anyone have experience with this? Obviously I can't remove the package path references because then all assets would fail in the consuming application.

**Comment by bartekpacia** [comment]
Hi @Calpoog, that's the first time I hear about such problem. Thanks for reporting.

You can get better logs when running with `patrol develop` - did you try it?

**Comment by Calpoog** [comment]
Totally missed that! But yes, that was much more helpful and confirmed that the first exception is indeed that an asset could not be found. As linked above, supposedly integration tests should work with package paths for assets in the same package, but it definitely does not appear to be the case here. I'm wondering if it's because Patrol has it's own custom bindings? I'm pretty busy but maybe I can set up the example app and add a single package reference asset to validate this.

**Reply by bartekpacia** [reply]
A simple reproducer app would certainly help!

**Comment by Calpoog** [comment]
I confirmed by modifying my simple test. Here is my test:

```dart
void main() {
  Future<Widget> setup() async {
    return Column(
      children: [
        Image.asset('assets/Images/Loading/loading-spinner.png'),
        Image.asset(
            'packages/my_package/assets/Images/Loading/loading-spinner.png'),
      ],
    );
  }

  patrolTest(
    'The app loads past the splash screen',
    ($) async {
      final app = await setup();
      await $.pumpWidgetAndSettle(app);

      expect(true, isTrue);
    },
  );
}
```

If I remove the second image, the test succeeds. If I leave it, the test fails. The test just tests true=true, so the uncaught asset exception is causing the failure. This package path works outside of this package in an app that consumes it.

**Reply by Calpoog** [reply]
Obviously while I'd like to see the tests support self-referential package asset paths, I found an alternative by checking my environment for a dart define. I have a some abstract classes used to retrieve paths for various assets at runtime, and I can use a conditional in there to change it from package paths to relative local paths. It would be nice if Patrol added a dart define for 'PATROL_TEST' or something that the app can use to make these kind of decisions.
